core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -4194,6 +4256,75 @@ private boolean rewriteNegativeLimitGuard(QueryModel model, SqlExecutionContext
                 );
     }
 
+    /**
+     * Handle queries like:
+     * SELECT timestamp, side FROM trades ORDER BY timestamp ASC, side DESC LIMIT -3
+     * This would ordinarily compile to a LimitedSizePartiallySortedRecordCursor.
+     * That means it will forward scan - and run out of memory on demo!
+     * Instead, we aim to subquery a backwards scan and then sort.
+     *
+     * @param model
+     */
+    private void rewriteNegativeLimitWithOrderBy(QueryModel model) throws SqlException {
+        if (model.getSelectModelType() == QueryModel.SELECT_MODEL_CHOOSE
+                && model.getNestedModel() != null
+                && model.getNestedModel().getSelectModelType() == QueryModel.SELECT_MODEL_NONE
+                && model.getNestedModel().getOrderBy() != null
+                && model.getNestedModel().getOrderBy().size() > 1
+                && model.getNestedModel().getTimestamp() != null
+                && model.getLimitLo() != null
+                && Chars.equals(model.getLimitLo().token, "-")) {
+
+            QueryModel nested = model.getNestedModel();
+
+            int firstOrderByDir = nested.getOrderByDirection().get(0);
+
+            if (firstOrderByDir != ORDER_DIRECTION_ASCENDING) {
+                return;
+            }
+
+            ExpressionNode firstOrderByArg = nested.getOrderBy().get(0);
+
+            if (!Chars.equals(firstOrderByArg.token, nested.getTimestamp().token)) {
+                return;
+            }
+
+            // we want to push down a limited reverse scan, and then sort into the intended ordering afterwards
+
+            // first, copy the order by up
+            for (int i = 0, n = nested.getOrderBy().size(); i < n; i++) {
+                model.addOrderBy(nested.getOrderBy().get(i), nested.getOrderByDirection().get(i));
+                model.getOrderByAdvice().add(nested.getOrderBy().get(i));
+                model.getOrderByDirectionAdvice().add(nested.getOrderByDirection().get(i));
+
+                // also strip advice
+                model.getOrderByAdvice().remove(i);
```

Comment:
What are we doing here? Are we removing older order by advice elements? If so, is it guaranteed that `model.getOrderByAdvice()`'s size matches the `nested.getOrderBy()`'s one?